src/math/exp_fcn.c

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Function Declarations */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b);
+
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = ((u64_rlo & BIT_MASK_LOW_Q27P5) != 0ULL);
+	u64_rlo =
+	    (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? (uint64_t)1 : (uint64_t)0);
+	return (int64_t)u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *bitshiptrout
+ *		  uint64_t *bitsloptrout
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout)
+{
+	uint64_t absIn0;
+	uint64_t absIn1;
+	uint64_t in0Hi;
+	uint64_t in0Lo;
+	uint64_t in1Hi;
+	uint64_t prodHiLo;
+	uint64_t prodLoHi;
+
+	absIn0 = (in_0 < 0LL) ? ~(uint64_t)in_0 + 1ULL : (uint64_t)in_0;
+	absIn1 = (in_1 < 0LL) ? ~(uint64_t)in_1 + 1ULL : (uint64_t)in_1;
+	in0Hi = absIn0 >> 32ULL;
+	in0Lo = absIn0 & UINT32_MAX;
+	in1Hi = absIn1 >> 32ULL;
+	absIn0 = absIn1 & UINT32_MAX;
+	prodHiLo = in0Hi * absIn0;
+	prodLoHi = in0Lo * in1Hi;
+	absIn0 *= in0Lo;
+	absIn1 = 0ULL;
+	in0Lo = absIn0 + (prodLoHi << 32ULL);
+	if (in0Lo < absIn0)
+		absIn1 = 1ULL;
+
+	absIn0 = in0Lo;
+	in0Lo += prodHiLo << 32ULL;
+	if (in0Lo < absIn0)
+		absIn1++;
+
+	absIn0 = ((absIn1 + in0Hi * in1Hi) + (prodLoHi >> 32ULL)) + (prodHiLo >> 32ULL);
+	if (in_0 != 0LL && (in_1 != 0LL && (in_0 > 0LL) != (in_1 > 0LL))) {
+		absIn0 = ~absIn0;
+		in0Lo = ~in0Lo;
+		in0Lo++;
+		if (in0Lo == 0ULL)
+			absIn0++;
+	}
+	*bitshiptrout = absIn0;
+	*bitsloptrout = in0Lo;
+}
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x    |  ts    |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |  1    | 32 | 23 |   0   | 4.28   | 9.23   |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t exp_int32(int32_t x)
```

Comment:
please specify the algorithm you used, is it taylor?
From performance perspective, please also hifi version together, c always have performance issue especially for such heavy computation, 2 - 3 times than asm verison.

Reply:
Sure, I'll include HiFi 3, 4, and 5 in my next pull request since the HiFi implementation will be based on the PR review at hand and I'll be include my modifications in the HiFi version.